src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,236 @@
+    for (ssize_t i=0; i<genome.aLinkGenes.size(); ++i) {
+      if (genome.aLinkGenes[i].FromNeuronId() == fromNeuronId &&
+          genome.aLinkGenes[i].ToNeuronId() == toNeuronId) {
+        return i;
+      }
+    }
+    return -1;  // -1 means not exist.
+  } 
+
+  // Mutate: add new link to genome.
+  void MutateAddLink(Genome& genome, double mutateAddLinkProb) {
+    // Whether mutate or not.
+    double p = mlpack::math::Random();
+    if (p > mutateAddLinkProb) return;
+
+    // Select from neuron
+    ssize_t fromNeuronIdx = mlpack::math::RandInt(0, genome.aNeuronGenes.size());
+    ssize_t fromNeuronId = genome.aNeuronGenes[fromNeuronIdx].Id();
+
+    // Select to neuron which cannot be input.
+    ssize_t toNeuronIdx = mlpack::math::RandInt(genome.NumInput(), genome.aNeuronGenes.size());
+    ssize_t toNeuronId = genome.aNeuronGenes[toNeuronIdx].Id();
+
+    // Check link already exist or not.
+    ssize_t linkIdx = IsLinkExist(genome, fromNeuronId, toNeuronId);
+    if (linkIdx != -1) {
+      genome.aLinkGenes[linkIdx].Enabled(true);
+      return;
+    }
+
+    // Check innovation already exist or not.
+    ssize_t innovIdx = CheckLinkInnovation(fromNeuronId, toNeuronId);
+    if (innovIdx != -1) {
+      LinkGene linkGene(fromNeuronId,
+                        toNeuronId,
+                        aLinkInnovations[innovIdx].newLinkInnovId,
+                        mlpack::math::RandNormal(0, 1), // TODO: make the distribution an argument for control?
+                        true);
+      genome.aLinkGenes.push_back(linkGene);
+      return;
+    }
+
+    // If new link and new innovation, create it, push new innovation.
+    LinkInnovation linkInnov = AddLinkInnovation(fromNeuronId, toNeuronId);
+    LinkGene linkGene(fromNeuronId,
+                      toNeuronId,
+                      linkInnov.newLinkInnovId,
+                      mlpack::math::RandNormal(0, 1), // TODO: make the distribution an argument for control?
+                      true);
+    genome.aLinkGenes.push_back(linkGene);
+  }
+
+  // Mutate: add new neuron to genome.
+  void MutateAddNeuron(Genome& genome, double mutateAddNeuronProb) {
+    // Whether mutate or not.
+    double p = mlpack::math::Random();
+    if (p > mutateAddNeuronProb) return;
+
+    // No link.
+    if (genome.aLinkGenes.size() == 0) return;
+
+    // Select link to split.
+    ssize_t linkIdx = mlpack::math::RandInt(0, genome.aLinkGenes.size());
+    if (!genome.aLinkGenes[linkIdx].Enabled()) return;
+
+    genome.aLinkGenes[linkIdx].Enabled(false);
+
+    // Check innovation already exist or not.
+    ssize_t splitLinkInnovId = genome.aLinkGenes[linkIdx].InnovationId();
+    ssize_t innovIdx = CheckNeuronInnovation(splitLinkInnovId);
+    if (innovIdx != -1) {
+      NeuronGene neuronGene(aNeuronInnovations[innovIdx].newNeuronId,
+                            HIDDEN,
+                            SIGMOID,  // TODO: make it random??
```

Comment:
NEAT doesn't change the activation function, but it could be interesting to see if that would increase the performance. However, we have to find a good way, to abstract that functionality from the rest of the code. I think, it's good for now to go with a static activation function.